asst2/CODE/my_malloc.c

```diff
@@ -4,39 +4,251 @@
 #include <stdlib.h>
 #include "my_malloc.h"
 
-char* MEM = NULL;
-
-struct sigaction* act;
+static char* myblock;
+static int firstMalloc = 1;
+struct sigaction segh;
+size_t page_size;  //size of page given to each process
+int num_pages; //number of pages available for user
+int num_segments; //number of segments within a page available for allocation 
+char* mem_space; //user space, after page table
 
 static void handler(int sig, siginfo_t* si, void* unused) {
-  printf("Got SIGSEGV at address: 0x%lx\n",(long) si->si_addr);
+	printf("Got SIGSEGV at address: 0x%lx\n",(long) si->si_addr);
+	exit(0);
+}
+
+//user blocks are occupied (malloc'd) if the
+//left most bit of 8 bit of the metadata is 1,
+//so to check if a user data block is occupied:
+//calculate bitwise '&' of size and 0x80 (0b 1000 0000)
+//if result is 0, then the block is free, 
+//and return this so is_occupied = false
+//else we will return 0x80 which is true 
+int is_occupied(metadata* curr) {
+	return  (curr->size) & 0x80;
+}
+
+//ignores the leftmost bit in the metadata which represents free/malloc
+//to return the number of userdata blocks
+unsigned char block_size(metadata* curr) {
+	return curr -> size & 0x7f;
```

Comment:
this is incorrect. 0x0 represents 1, not 0. We need to add 1, otherwise everything breaks.

Reply:
unless you've already factored this in elsewhere, in which case we'll be fine

Reply:
actually, I'm tripping. ignore this whole thread ^
I need sleep

Reply:
although we may want to revisit this in the future. If we can allocate more than a page at a time, we should have a flag that tells us that the allocation is more than a page. Then we'll have only 6 bits per metadata to store segment length. Then we'd have to do this normalization technique.

Reply:
yeah, there might have to be several metadata / pagedata changes to account for multiple page allocations, might have to discuss em